geps/gep-1897.md

```diff
@@ -125,7 +139,320 @@ This GEP is the outcome of the TLS use cases #4 and #5 in
+
+```go
+// TLSConnectionPolicy provides a way to publish TLS configuration
+// that enables a gateway client to connect to a backend pod.
+type TLSConnectionPolicy struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the desired state of TLSConnectionPolicy.
+    Spec TLSConnectionPolicySpec `json:"spec"`
+
+    // Status defines the current state of TLSConnectionPolicy.
+    Status TLSConnectionPolicyStatus `json:"status,omitempty"`
+}
+
+// TLSConnectionPolicySpec defines the desired state of
+// TLSConnectionPolicy.
+// Note: there is no Override or Default policy configuration.
+//
+// Support: Core
+type TLSConnectionPolicySpec struct {
+    // TargetRef identifies an API object to apply policy to.
+    // Services are the only valid API target references.
+    // Note that this config applies to the entire referenced resource
+    // by default, but this default may change in the future to provide
+    // a more granular application of the policy.
+    TargetRef gatewayv1a2.PolicyTargetReference `json:"targetRef"`
+
+    // TLS contains TLS connection policy configuration.
+    TLS *TLSConnectionPolicyConfig `json:”tls”`
+}
+
+// TLSConnectionPolicyConfig contains TLS connection policy configuration.
+type TLSConnectionPolicyConfig struct {
+    // TrustedCACertRefs contains one or more references to
+    // Kubernetes objects that contain PEM-encoded TLS certificates,
+    // which are used to establish a TLS handshake between the gateway
+    // and backend pod.
+    //
+    // If the TrustedCACertRefs is empty, then the system trusted
+    // certificates should be used. If there are none, or the
+    // implementation doesn't define system trusted certificates,
+    // then a TLS connection must fail.
+    //
+    // A single TrustedCACertRef to a Kubernetes ConfigMap has "Core"
+    // support.  Implementations MAY choose to support attaching
+    // multiple certificates to a backend, but this behavior is
+    // implementation-specific.
+    //
+    // References to a resource in a different namespace are 
+    // invalid.
+    // 
+    // Support: Core - An optional single reference to a Kubernetes
+    // ConfigMapName.
+    //
+    // Support: Implementation-specific (No reference, more than one
+    // reference, or other resource types.)
+    //
+    // +kubebuilder:validation:MaxItems=64
+    // +optional
+    TrustedCACertRefs []ConfigMapName `json:”trustedCACertRefs,omitempty”`
+
+    // AllowedSANs are the Subject Alternative Names that are allowed for
+    // this TLS connection.  It represents the hostnames that may
+    // authenticate, and must be in the received certificate.
+    //
+    // If AllowedSANs is set, then an implementation must validate that
+    // the certificate provided by the pod represents a SAN in its list.
+    //
+    // If AllowedSANs is empty, then any SAN is allowed.
```

Comment:
@howardjohn @keithmattix - I'm adding the `backendCABundle` that Nick mentioned in https://github.com/kubernetes-sigs/gateway-api/pull/2113#issuecomment-1651028525.   Could this be used if the
AllowedSANs field is absent, and the use case was self-signed certs?

Reply:
In TLS we ought to either verify the SAN _or_ make a user explicitly opt into "Insecure Mode".

So I disagree with Nick's suggestion. `backendCABundle` is insecure-by-design and cannot be accepted IMO. TLS _must_ verify certificate fully: that means the CA but also properties of the cert, most notably the SANs.

Reply:
Or follow the standards - SNI, SAN and authority set to the hostname of the backend - and verify standard DNS cert. 

SAN is required for spiffee or custom schemes. I don't think the standard DNS case needs anything special.

Reply:
I think the ca bundle and SAN verification are orthogonal pieces of the same process. SAN verification says "even if we have the same trust chain, I need to know that I'm talking to the identity I'm intending to"

Reply:
@costinm just to be clear, are you talking about standard Server Name Indication (https://datatracker.ietf.org/doc/html/rfc6066#section-3)   where it states, "Clients, however, SHOULD
   store the SNI with the PSK to fulfill the requirements of
   [Section 4.6.1](https://datatracker.ietf.org/doc/html/rfc8446#section-4.6.1)." ?  It would be helpful if you could point out the standard you're thinking we should follow.

Reply:
@howardjohn I removed that late addition, and for this iteration, TLS will just fail if no `AllowedSANs` are specified.  In future iterations we could enhance the behavior once we come to consensus on what we want.